### PR TITLE
[CARBONDATA-3754]avoid listing index files during SI rebuild

### PR DESCRIPTION
**Pull request by akashrn5**
### Why is this PR needed?
 List files was done during rebuild of SI, which will be costly in case of S3 and OBS
 
 ### What changes were proposed in this PR?
avoided list files and delete files handled through segmentFileStore
    
 ### Does this PR introduce any user interface change?
 - No
 - Yes. (please explain the change and update document)

 ### Is any new testcase added?
 - No(existing tests will takecare)

### CODE REVIEW

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala [hunk 1]
```diff
@@ -280,6 +284,40 @@ object SecondaryIndexUtil {
...
   }
 
+  /**
+   * This method deletes the old index files or merge index file after data files merge
+   */
+  def deleteOldIndexOrMergeIndexFiles(carbonLoadModel: CarbonLoadModel,
+      validSegments: util.List[Segment],
+      indexCarbonTable: CarbonTable): Unit = {
+    // delete the index/merge index carbonFile of old data files
+    val loadModel = SecondaryIndexUtil
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala:294]: Why LoadModel is formed again?, as we already have carbonLoadModel in methodParams

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala [hunk 2]
```diff
@@ -280,6 +284,40 @@ object SecondaryIndexUtil {
...
+        val segmentPath = CarbonTablePath.getSegmentPath(indexCarbonTable.getTablePath,
+          segment.getSegmentNo)
+        indexFiles = new SegmentIndexFileStore().getMergeOrIndexFilesFromSegment(segmentPath)
+      } else {
+        val segmentFileStore = new SegmentFileStore(indexCarbonTable.getTablePath,
+          segment.getSegmentFileName)
+        indexFiles = segmentFileStore.getIndexOrMergeFiles
+      }
+    }
+    val indexFilesToDelete = indexFiles.keySet().asScala.filter { indexFile =>
```
Comment (Indhumathi27) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala:312]: I think for each segment we have to delete old indexFiles. This delete code has to be moved inside foreach loop?

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1051/

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala (hunk 1 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: you are right, removed

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala (hunk 2 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: yes

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1054/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2767/

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala [hunk 3]
```diff
@@ -280,6 +286,35 @@ object SecondaryIndexUtil {
     }
   }
 
+  /**
+   * This method deletes the old index files or merge index file after data files merge
+   */
+  def deleteOldIndexOrMergeIndexFiles(factTimeStamp: Long,
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala:292]: make it private and move factTimeStamp to next line

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala [hunk 4]
```diff
@@ -280,6 +286,35 @@ object SecondaryIndexUtil {
...
+      } else {
+        val segmentFileStore = new SegmentFileStore(indexCarbonTable.getTablePath,
+          segment.getSegmentFileName)
+        indexFiles = segmentFileStore.getIndexOrMergeFiles
+      }
+      val indexFilesToDelete = indexFiles.keySet().asScala.filter { indexFile =>
+        DataFileUtil.getTimeStampFromFileName(indexFile).toLong >
+        factTimeStamp
+      }
+      indexFilesToDelete.foreach { indexFile =>
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala:312]: no need to declare local variable indexFilesToDelete

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala [hunk 5]
```diff
@@ -280,6 +286,35 @@ object SecondaryIndexUtil {
...
+   * This method deletes the old index files or merge index file after data files merge
+   */
+  def deleteOldIndexOrMergeIndexFiles(factTimeStamp: Long,
+      validSegments: util.List[Segment],
+      indexCarbonTable: CarbonTable): Unit = {
+    // delete the index/merge index carbonFile of old data files
+    validSegments.asScala.foreach { segment =>
+      var indexFiles: util.Map[String, String] = null
+      // for old store scenario
+      if (segment.getSegmentFileName == null) {
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala:299]: move line 299 to 307 as a function in CarbonTable to return the index file list for specified segment list

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala [hunk 6]
```diff
@@ -197,6 +200,9 @@ object SecondaryIndexUtil {
       }
       if (finalMergeStatus) {
         if (null != mergeStatus && mergeStatus.length != 0) {
+          deleteOldIndexOrMergeIndexFiles(carbonLoadModel.getFactTimeStamp,
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala:203]: move carbonLoadModel.getFactTimeStamp to  next line

Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: done

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala (hunk 5 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: done

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala (hunk 4 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: done

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala (hunk 3 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1066/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2779/

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala [hunk 7]
```diff
@@ -280,6 +286,23 @@ object SecondaryIndexUtil {
...
+   */
+  private def deleteOldIndexOrMergeIndexFiles(
+      factTimeStamp: Long,
+      validSegments: util.List[Segment],
+      indexCarbonTable: CarbonTable): Unit = {
+    // delete the index/merge index carbonFile of old data files
+    validSegments.asScala.foreach { segment =>
+      indexCarbonTable.getIndexFilesListForSegment(segment).asScala.filter { indexFile =>
+        DataFileUtil.getTimeStampFromFileName(indexFile).toLong > factTimeStamp
+      }.foreach { indexFile =>
```
Comment (kunal642) [inline integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala:300]: combine filter and foreach into a single foreach

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 8]
```diff
@@ -1309,4 +1313,22 @@ public String getParentTableName() {
     return allIndexes;
   }
 
+  /**
+   * This method returns the list of indx/merge index files for a segment in carbonTable.
+   */
+  public Set<String> getIndexFilesListForSegment(Segment segment) throws IOException {
```
Comment (kunal642) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java:1319]: I think this method belongs in SegmentFileStore.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2798/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1086/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2818/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1105/

File: integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala (hunk 7 above)
Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/secondaryindex/util/SecondaryIndexUtil.scala]: done

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java (hunk 8 above)
Comment (akashrn5) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: done

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**